### PR TITLE
Move caching of values out of providers

### PR DESCRIPTION
**Pull request by axelboc**
As per the roadmap in #417, the caching of dataset values is now done in hooks `useDatasetValue` and `useDatasetValues`. The `values` cache itself is a `Map` located in `Provider.ts`.

**EDIT** the caching of the values is still out of the providers, but it is now handled solely by our suspense library, [react-suspense-fetch](https://github.com/dai-shi/react-suspense-fetch). Until now, we used `use-async-resource`, which already had its own internal cache but was tricky to use and inefficient, notably because it relied on a hook (one hook declaration = one cache, which prevented us from having multiple components/hooks share the same cache).

---

@loichuder I'll merge my PRs into branch `fetching-strategy-2` and I'll create a final PR for it when I'm done. This will allow me to keep merging while still allowing you to review everything when you're back.

### CODE REVIEW

**Comment by axelboc** [comment]
Can't get around this regression in the tests...

Also, I'm not happy with the way our cache interacts with `useAsyncResource`, which also has its own cache: `useAsyncResource(asyncFn, params)` starts fetching right away, before we have the opportunity to check our own cache, so it feels wrong.

- We could just use `useAsyncResource`'s cache alone, but it wouldn't be shared between the two `useDatasetValue(s)` hooks, so it's not great...
- If I use the hook lazily with `useAsyncResource(asyncFn)`, then `valueReader()` can return `undefined`, which means we have to handle `undefined` everywhere.

Basically, I feel like `useAsyncResource` does more than we need, and the fact that it's a hook really limits the way we can use it. I'm going to look for a lower-level `Suspense` library

File: src/h5web/providers/Provider.tsx [hunk 1]
```diff
@@ -12,32 +14,29 @@ function Provider(props: Props): ReactElement {
...
-    api,
-  ]);
+  const { value: metadata, error } = useAsync(async () => {
+    return api.fetchMetadata();
+  }, [api]);
+
+  const valuesStore = useMemo(
+    () => createFetchStore(api.fetchValue.bind(api)),
+    [api]
+  );
```
Comment (axelboc) [inline src/h5web/providers/Provider.tsx:24]: Creating the values store here, with the provider API's `fetchValue` async method. The store has an internal cache that matches the signature of `fetchValue` (i.e. an object with `HDF5Id` keys and `HDF5Value` values).

File: src/h5web/providers/jupyter/api.ts [hunk 2]
```diff
@@ -60,7 +54,7 @@ export class JupyterApi implements ProviderAPI {
     return data;
   }
 
-  private async fetchMetadata(path: string): Promise<JupyterMetaResponse> {
+  private async _fetchMetadata(path: string): Promise<JupyterMetaResponse> {
```
Comment (loichuder) [inline src/h5web/providers/jupyter/api.ts:57]: Could be renamed `fetchEntityMetadata`

Comment (axelboc) [reply src/h5web/providers/jupyter/api.ts:57]: FYI, I've renamed it back to `fetchMetadata`, but yeah, your suggestion is more explicit.
